ape_hardhat/__init__.py

```diff
@@ -0,0 +1,237 @@
+    fork_url: Optional[str] = None
+
+    # --fork-block-number <INT, block number to fork from>
+    fork_block_number: Optional[int] = None
+
+    # --port <INT, default from Hardhat is 8545, but our default is to assign a random port number>
+    port: Optional[int] = None
+
+    # retry strategy configs, try increasing these if you're getting HardhatSubprocessError
+    network_retries: List[float] = HARDHAT_START_NETWORK_RETRIES
+    process_attempts: int = HARDHAT_START_PROCESS_ATTEMPTS
+
+
+class HardhatProvider(EthereumProvider):
+    def __post_init__(self):
+        self._hardhat_web3 = (
+            None  # we need to maintain a separate per-instance web3 client for Hardhat
+        )
+        self.port = None
+        self.process = None
+
+        hardhat_config_file = self.network.config_manager.PROJECT_FOLDER / "hardhat.config.js"
+
+        # Write the hardhat config file in the Ape project dir if it doesn't exist yet
+        if not hardhat_config_file.is_file():
+            hardhat_config_file.write_text(HARDHAT_CONFIG)
+
+        # Check if npx and hardhat are installed
+        if call(["npx", "--version"], stderr=PIPE, stdout=PIPE, stdin=PIPE) != 0:
+            raise HardhatSubprocessError(
+                "Missing npx binary. See ape-hardhat README for install steps."
+            )
+        if call(["npx", "hardhat", "--version"], stderr=PIPE, stdout=PIPE, stdin=PIPE) != 0:
+            raise HardhatSubprocessError(
+                "Missing hardhat NPM package. See ape-hardhat README for install steps."
+            )
+
+        # register atexit handler to make sure disconnect is called for normal object lifecycle
+        atexit.register(self.disconnect)
+
+        # register signal handlers to make sure atexit handlers are called when the parent python
+        # process is killed
+        signal.signal(signal.SIGINT, _signal_handler)
+        signal.signal(signal.SIGTERM, _signal_handler)
+
+    def _start_process(self):
+        """Start the hardhat process and wait for it to respond over the network."""
+        # handle configs
+        cmd = ["npx", "hardhat", "node"]
+        if self.config.fork_url:
+            cmd.extend(["--fork", self.config.fork_url])
+        if self.config.fork_block_number:
+            cmd.extend(["--fork-block-number", str(self.config.fork_block_number)])
+
+        # pick a random port if one isn't configured
+        self.port = self.config.port
+        if not self.port:
+            self.port = random.randint(EPHEMERAL_PORTS_START, EPHEMERAL_PORTS_END)
+        cmd.extend(["--port", str(self.port)])
+
+        # TODO: Add configs to send stdout to logger / redirect to a file in plugin data dir?
+        process = Popen(cmd, stdin=PIPE, stdout=PIPE, stderr=PIPE, preexec_fn=_set_death_signal)
+        connected = False
+        for retry_time in self.config.network_retries:
+            time.sleep(retry_time)
+            super().connect()
+            try:
```

Comment:
This whole `try` block might be a good candidate for a helper method, e.g. `_verify_connection()` and then just have `connected = `_verify_connect()`.

This whole `_start_process()` method is getting kind of lengthy, so I was just looking for ways to help break it up

Reply:
Good idea, I moved that block of code into a `_verify_connection` method.